File: backend/src/routes/catalog.js
const express = require('express');
const { body, query, validationResult } = require('express-validator');
const { Package, PricingRule, AccessZone } = require('../models/catalog');
const { authMiddleware } = require('../middleware/auth');
const logger = require('../utils/logger');

const router = express.Router();

// Get packages for a store
router.get('/packages', [
  authMiddleware,
  query('store_id').optional().isString(),
  query('type').optional().isIn(['single', 'multi', 'timepass', 'credit', 'bundle']),
  query('active').optional().isBoolean()
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'E_VALIDATION',
        message: 'Validation failed',
        details: errors.array()
      });
    }

    const { type, active } = req.query;
    const store_id = req.query.store_id || req.store_id;

    // Build query
    const query = {
      tenant_id: req.tenant_id,
      store_id: store_id
    };

    if (type) query.type = type;
    if (active !== undefined) query.active = active === 'true';

    const packages = await Package.find(query)
      .sort({ name: 1 });

    // Filter by active window if needed
    const filteredPackages = packages.filter(pkg => {
      if (!pkg.active) return false;
      
      if (pkg.active_window && pkg.active_window.from && pkg.active_window.to) {
        const now = new Date();
        const currentTime = now.getHours() * 60 + now.getMinutes();
        const fromTime = pkg.parseTime(pkg.active_window.from);
        const toTime = pkg.parseTime(pkg.active_window.to);
        
        return currentTime >= fromTime && currentTime <= toTime;
      }
      
      return true;
    });

    res.json({
      packages: filteredPackages.map(pkg => ({
        package_id: pkg.package_id,
        name: pkg.name,
        type: pkg.type,
        price: pkg.price,
        price_text: pkg.price_text,
        quota_or_minutes: pkg.quota_or_minutes,
        description: pkg.description,
        image_url: pkg.image_url,
        allowed_devices: pkg.allowed_devices,
        visible_on: pkg.visible_on,
        active_window: pkg.active_window
      }))
    });

  } catch (error) {
    next(error);
  }
});

// Get pricing rules for a store
router.get('/pricing/rules', [
  authMiddleware,
  query('store_id').optional().isString(),
  query('active').optional().isBoolean()
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'E_VALIDATION',
        message: 'Validation failed',
        details: errors.array()
      });
    }

    const { active } = req.query;
    const store_id = req.query.store_id || req.store_id;

    // Build query for both store and tenant rules
    const query = {
      $or: [
        { scope: 'store', scope_id: store_id },
        { scope: 'tenant', scope_id: req.tenant_id }
      ]
    };

    if (active !== undefined) query.active = active === 'true';

    const rules = await PricingRule.find(query)
      .sort({ priority: -1, created_at: 1 });

    // Filter by active window
    const now = new Date();
    const activeRules = rules.filter(rule => {
      if (!rule.active) return false;
      
      if (rule.active_window) {
        const from = rule.active_window.from;
        const to = rule.active_window.to;
        
        if (from && to) {
          return now >= from && now <= to;
        }
      }
      
      return true;
    });

    res.json({
      rules: activeRules.map(rule => ({
        rule_id: rule.rule_id,
        name: rule.name,
        description: rule.description,
        kind: rule.kind,
        scope: rule.scope,
        scope_id: rule.scope_id,
        params: rule.params,
        priority: rule.priority,
        active_window: rule.active_window
      }))
    });

  } catch (error) {
    next(error);
  }
});

// Calculate pricing for a cart
router.post('/pricing/calculate', [
  authMiddleware,
  body('items').isArray().withMessage('Items array is required'),
  body('items.*.package_id').notEmpty().withMessage('Package ID is required'),
  body('items.*.quantity').isInt({ min: 1 }).withMessage('Quantity must be at least 1')
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'E_VALIDATION',
        message: 'Validation failed',
        details: errors.array()
      });
    }

    const { items, promo_code } = req.body;

    // Get packages
    const packageIds = items.map(item => item.package_id);
    const packages = await Package.find({
      package_id: { $in: packageIds },
      tenant_id: req.tenant_id,
      store_id: req.store_id,
      active: true
    });

    const packageMap = {};
    packages.forEach(pkg => {
      packageMap[pkg.package_id] = pkg;
    });

    // Get pricing rules
    const rules = await PricingRule.find({
      $or: [
        { scope: 'store', scope_id: req.store_id },
        { scope: 'tenant', scope_id: req.tenant_id }
      ],
      active: true
    }).sort({ priority: -1 });

    // Calculate line items
    const lineItems = items.map(item => {
      const pkg = packageMap[item.package_id];
      if (!pkg) {
        throw new Error(`Package ${item.package_id} not found`);
      }

      const lineTotal = pkg.price * item.quantity;
      const discounts = [];

      // Apply line-level discounts
      rules.forEach(rule => {
        if (rule.kind === 'line_percent' && 
            (!rule.params.target_packages || rule.params.target_packages.includes(item.package_id))) {
          const discountAmount = lineTotal * (rule.params.percentage / 100);
          discounts.push({
            type: 'percent',
            value: rule.params.percentage,
            amount: discountAmount,
            description: rule.name,
            rule_id: rule.rule_id
          });
        } else if (rule.kind === 'line_amount' && 
                   (!rule.params.target_packages || rule.params.target_packages.includes(item.package_id))) {
          const discountAmount = Math.min(rule.params.amount, lineTotal);
          discounts.push({
            type: 'amount',
            value: rule.params.amount,
            amount: discountAmount,
            description: rule.name,
            rule_id: rule.rule_id
          });
        }
      });

      const lineDiscountTotal = discounts.reduce((sum, discount) => sum + discount.amount, 0);
      const lineSubtotal = lineTotal - lineDiscountTotal;

      return {
        package_id: item.package_id,
        package_name: pkg.name,
        quantity: item.quantity,
        unit_price: pkg.price,
        line_total: lineTotal,
        discounts: discounts,
        line_discount_total: lineDiscountTotal,
        line_subtotal: lineSubtotal
      };
    });

    // Calculate cart totals
    const subtotal = lineItems.reduce((sum, item) => sum + item.line_subtotal, 0);
    const lineDiscountTotal = lineItems.reduce((sum, item) => sum + item.line_discount_total, 0);

    // Apply cart-level discounts
    const cartDiscounts = [];
    let cartDiscountTotal = 0;

    rules.forEach(rule => {
      if (rule.kind === 'cart_percent') {
        const discountAmount = subtotal * (rule.params.percentage / 100);
        cartDiscounts.push({
          type: 'percent',
          value: rule.params.percentage,
          amount: discountAmount,
          description: rule.name,
          rule_id: rule.rule_id
        });
        cartDiscountTotal += discountAmount;
      } else if (rule.kind === 'cart_amount') {
        const discountAmount = Math.min(rule.params.amount, subtotal);
        cartDiscounts.push({
          type: 'amount',
          value: rule.params.amount,
          amount: discountAmount,
          description: rule.name,
          rule_id: rule.rule_id
        });
        cartDiscountTotal += discountAmount;
      } else if (rule.kind === 'promo_code' && promo_code === rule.params.promo_code) {
        const discountAmount = rule.params.amount || (subtotal * (rule.params.percentage / 100));
        cartDiscounts.push({
          type: 'promo_code',
          value: rule.params.amount || rule.params.percentage,
          amount: discountAmount,
          description: rule.name,
          rule_id: rule.rule_id
        });
        cartDiscountTotal += discountAmount;
      }
    });

    const finalSubtotal = subtotal - cartDiscountTotal;
    const taxTotal = finalSubtotal * 0.07; // 7% VAT (configurable)
    const grandTotal = finalSubtotal + taxTotal;

    res.json({
      line_items: lineItems,
      subtotal: subtotal,
      line_discount_total: lineDiscountTotal,
      cart_discounts: cartDiscounts,
      cart_discount_total: cartDiscountTotal,
      final_subtotal: finalSubtotal,
      tax_total: taxTotal,
      grand_total: grandTotal,
      currency: 'THB'
    });

  } catch (error) {
    next(error);
  }
});

// Get access zones
router.get('/zones', [
  authMiddleware,
  query('store_id').optional().isString()
], async (req, res, next) => {
  try {
    const store_id = req.query.store_id || req.store_id;

    const zones = await AccessZone.find({
      tenant_id: req.tenant_id,
      store_id: store_id,
      active: true
    }).populate('devices packages');

    res.json({
      zones: zones.map(zone => ({
        zone_id: zone.zone_id,
        name: zone.name,
        description: zone.description,
        devices: zone.devices,
        packages: zone.packages
      }))
    });

  } catch (error) {
    next(error);
  }
});

module.exports = router;

File: backend-fastapi/app/models/catalog.py
"""
Catalog Models
"""
from datetime import datetime
from typing import Optional, List, Dict, Any, Union
from pydantic import BaseModel, Field
from enum import Enum

from .core import BaseDocument


class PackageType(str, Enum):
    """Package types"""
    SINGLE = "single"
    MULTI = "multi"
    TIME_BASED = "time_based"


class PricingRuleKind(str, Enum):
    """Pricing rule kinds"""
    LINE_PERCENT = "line_percent"
    LINE_AMOUNT = "line_amount"
    CART_PERCENT = "cart_percent"
    CART_AMOUNT = "cart_amount"
    BULK_DISCOUNT = "bulk_discount"


class AccessZone(BaseDocument):
    """Access zone model"""
    
    zone_id: str = Field(..., description="Unique zone identifier")
    tenant_id: str = Field(..., description="Parent tenant ID")
    store_id: str = Field(..., description="Parent store ID")
    name: str = Field(..., description="Zone name")
    description: Optional[str] = Field(default=None, description="Zone description")
    capacity: Optional[int] = Field(default=None, description="Zone capacity")
    status: str = Field(default="active", description="Zone status")
    settings: Dict[str, Any] = Field(default_factory=dict, description="Zone settings")
    
    class Config:
        collection = "access_zones"


class Package(BaseDocument):
    """Package model"""
    
    package_id: str = Field(..., description="Unique package identifier")
    tenant_id: str = Field(..., description="Parent tenant ID")
    store_id: str = Field(..., description="Parent store ID")
    name: str = Field(..., description="Package name")
    description: Optional[str] = Field(default=None, description="Package description")
    type: PackageType = Field(..., description="Package type")
    price: float = Field(..., ge=0, description="Package price")
    quota_or_minutes: int = Field(..., ge=1, description="Quota or minutes")
    active: bool = Field(default=True, description="Package active status")
    visible_on: List[str] = Field(default_factory=list, description="Visible on devices")
    access_zones: List[str] = Field(default_factory=list, description="Access zones")
    settings: Dict[str, Any] = Field(default_factory=dict, description="Package settings")
    
    class Config:
        collection = "packages"


class PricingRule(BaseDocument):
    """Pricing rule model"""
    
    rule_id: str = Field(..., description="Unique rule identifier")
    tenant_id: str = Field(..., description="Parent tenant ID")
    store_id: str = Field(..., description="Parent store ID")
    name: str = Field(..., description="Rule name")
    description: Optional[str] = Field(default=None, description="Rule description")
    kind: PricingRuleKind = Field(..., description="Rule kind")
    params: Dict[str, Any] = Field(..., description="Rule parameters")
    priority: int = Field(default=0, description="Rule priority")
    active: bool = Field(default=True, description="Rule active status")
    conditions: Dict[str, Any] = Field(default_factory=dict, description="Rule conditions")
    settings: Dict[str, Any] = Field(default_factory=dict, description="Rule settings")
    
    class Config:
        collection = "pricing_rules"


class PricingCalculationRequest(BaseModel):
    """Pricing calculation request"""
    
    items: List[Dict[str, Any]] = Field(..., description="Items to calculate pricing for")
    customer_id: Optional[str] = Field(default=None, description="Customer ID")
    store_id: str = Field(..., description="Store ID")


class PricingCalculationResponse(BaseModel):
    """Pricing calculation response"""
    
    subtotal: float = Field(..., description="Subtotal amount")
    discount_total: float = Field(..., description="Total discount amount")
    tax_total: float = Field(..., description="Total tax amount")
    grand_total: float = Field(..., description="Grand total amount")
    applied_rules: List[Dict[str, Any]] = Field(default_factory=list, description="Applied rules")
    breakdown: Dict[str, Any] = Field(default_factory=dict, description="Price breakdown")


class Category(BaseDocument):
    """Category model"""
    
    category_id: str = Field(..., description="Unique category identifier")
    tenant_id: str = Field(..., description="Parent tenant ID")
    store_id: str = Field(..., description="Parent store ID")
    name: str = Field(..., description="Category name")
    description: Optional[str] = Field(default=None, description="Category description")
    parent_id: Optional[str] = Field(default=None, description="Parent category ID")
    sort_order: int = Field(default=0, description="Sort order")
    active: bool = Field(default=True, description="Category active status")
    settings: Dict[str, Any] = Field(default_factory=dict, description="Category settings")
    
    class Config:
        collection = "categories"


class Product(BaseDocument):
    """Product model"""
    
    product_id: str = Field(..., description="Unique product identifier")
    tenant_id: str = Field(..., description="Parent tenant ID")
    store_id: str = Field(..., description="Parent store ID")
    category_id: Optional[str] = Field(default=None, description="Category ID")
    name: str = Field(..., description="Product name")
    description: Optional[str] = Field(default=None, description="Product description")
    sku: Optional[str] = Field(default=None, description="Product SKU")
    price: float = Field(..., ge=0, description="Product price")
    cost: Optional[float] = Field(default=None, ge=0, description="Product cost")
    active: bool = Field(default=True, description="Product active status")
    settings: Dict[str, Any] = Field(default_factory=dict, description="Product settings")
    
    class Config:
        collection = "products"
